File: mesop/components/card/card.py
from typing import Literal

import mesop.components.card.card_pb2 as card_pb
from mesop.component_helpers import (
  Style,
  insert_composite_component,
  register_native_component,
)


@register_native_component
def card(
  *,
  appearance: Literal["outlined", "raised"] = "outlined",
  style: Style | None = None,
  key: str | None = None,
):
  """
  This function creates a card.

  Args:
      appearance: Card appearance style: outlined or raised.
      style: Style for the component.
      key: The component [key](../components/index.md#component-key).
  """
  return insert_composite_component(
    key=key,
    type_name="card",
    style=style,
    proto=card_pb.CardType(
      appearance=appearance,
    ),
  )

File: mesop/components/card/card.ts
import {MatCardModule} from '@angular/material/card';
import {Component, Input} from '@angular/core';
import {
  Key,
  Type,
  Style,
} from 'mesop/mesop/protos/ui_jspb_proto_pb/mesop/protos/ui_pb';
import {CardType} from 'mesop/mesop/components/card/card_jspb_proto_pb/mesop/components/card/card_pb';
import {formatStyle} from '../../web/src/utils/styles';

@Component({
  selector: 'mesop-card',
  templateUrl: 'card.ng.html',
  standalone: true,
  imports: [MatCardModule],
})
export class CardComponent {
  @Input({required: true}) type!: Type;
  @Input() key!: Key;
  @Input() style!: Style;
  private _config!: CardType;

  ngOnChanges() {
    this._config = CardType.deserializeBinary(
      this.type.getValue() as unknown as Uint8Array,
    );
  }

  config(): CardType {
    return this._config;
  }

  getStyle(): string {
    return formatStyle(this.style);
  }

  getAppearance(): 'outlined' | 'raised' {
    return this.config().getAppearance() as 'outlined' | 'raised';
  }
}
